collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java

```diff
@@ -0,0 +1,248 @@
+        return getIDOfFolderWithName(GOOGLE_DRIVE_SUBFOLDER, rootFolderId, true);
+    }
+
+    /**
+     * Searches for the folder saved in google drive with the given name and returns it's id
+     *
+     * @param name                   The name of the folder saved in google drive
+     * @param inFolder               The id of the folder containing the given folder
+     * @param shouldCreateIfNotFound If the given folder is not found then this parameter decides
+     *                               whether to create a new folder with the given name or not
+     */
+    @Nullable
+    public String getIDOfFolderWithName(@NonNull String name, @Nullable String inFolder, boolean shouldCreateIfNotFound)
+            throws IOException, MultipleFoldersFoundException {
+
+        String id = null;
+
+        // check if the folder exists
+        List<com.google.api.services.drive.model.File> files = getFilesFromDrive(name, inFolder);
+
+        if (files.size() > 1) {
+            throw new MultipleFoldersFoundException("Multiple \"" + name + "\" folders found");
+        }
+
+        if (files.size() == 1) {
+            id = files.get(0).getId();
+        } else if (files.size() == 0 && shouldCreateIfNotFound) {
+            id = createFolderInDrive(name, inFolder).getId();
+        }
+
+        return id;
+    }
+
+    /**
+     * Upload a file to google drive
+     *
+     * @param mediaName           The name of the uploaded file
+     * @param destinationFolderID Id of the folder into which the file has to be uploaded
+     * @param toUpload            The file which is to be uploaded
+     * @return id of the file if uploaded successfully
+     */
+    public String uploadFileToDrive(String mediaName, String destinationFolderID, File toUpload)
+            throws IOException {
+
+        //adding meta-data to the file
+        com.google.api.services.drive.model.File fileMetadata =
+                new com.google.api.services.drive.model.File()
+                        .setName(mediaName)
+                        .setViewersCanCopyContent(true)
+                        .setParents(Collections.singletonList(destinationFolderID));
+
+        String type = FileUtils.getMimeType(toUpload.getPath());
+        FileContent mediaContent = new FileContent(type, toUpload);
+
+        return drive.files()
+                .create(fileMetadata, mediaContent)
+                .setFields("id, parents")
+                .setIgnoreDefaultVisibility(true)
+                .execute()
+                .getId();
+    }
+
+    /**
+     * Creates a new folder in google drive
+     *
+     * @param folderName The name of the new folder
+     * @param parentId   The id of the folder in which we want to create the new folder
+     * @return the folder object if created successfully
+     */
+    private com.google.api.services.drive.model.File createFolderInDrive(@NonNull String folderName,
```

Comment:
This method does three separate things: 1. Create and prepare a new File Metadata, 2. Make a Drive call to create a Folder, and 3. Make a Drive call to set permissions. These should probably be three different methods.